atlasdb-config/src/main/java/com/palantir/atlasdb/factory/InitialisingTransactionManager.java

```diff
@@ -0,0 +1,630 @@
+    private final InitialisingTimestampService timestampService = InitialisingTimestampService.createUninitialised();
+    private final InitialisingPersistentLockService persistentLockService = InitialisingPersistentLockService.create();
+    private final InitialisingSweeperService sweeperService = InitialisingSweeperService.create();
+
+    private final ServiceDiscoveringAtlasSupplier atlasFactory;
+    private final AtlasDbConfig config;
+    private final java.util.function.Supplier<java.util.Optional<AtlasDbRuntimeConfig>> runtimeConfigSupplier;
+    private final Set<Schema> schemas;
+    private final TransactionManagers.Environment env;
+    private final LockServerOptions lockServerOptions;
+    private final boolean allowHiddenTableAccess;
+    private final String userAgent;
+
+    static Consumer<Runnable> runAsync = task -> {
+        Thread thread = new Thread(task);
+        thread.setDaemon(true);
+        thread.start();
+    };
+
+    public InitialisingTransactionManager(
+            AtlasDbConfig config,
+            java.util.function.Supplier<java.util.Optional<AtlasDbRuntimeConfig>> runtimeConfigSupplier,
+            Set<Schema> schemas,
+            TransactionManagers.Environment env,
+            LockServerOptions lockServerOptions,
+            boolean allowHiddenTableAccess,
+            String userAgent) {
+
+        atlasFactory = new ServiceDiscoveringAtlasSupplier(config);
+        this.config = config;
+        this.runtimeConfigSupplier = runtimeConfigSupplier;
+        this.schemas = schemas;
+        this.env = env;
+        this.lockServerOptions = lockServerOptions;
+        this.allowHiddenTableAccess = allowHiddenTableAccess;
+        this.userAgent = userAgent;
+
+        LockRequest.setDefaultLockTimeout(
+                SimpleTimeDuration.of(config.getDefaultLockTimeoutSeconds(), TimeUnit.SECONDS));
+
+        registerEndpoints();
+
+        try {
+            initialise();
+        } catch (Throwable th) {
+            log.info("Synchronous initialisation failed, initialisation will be done asynchronously", th);
+            initialiseAsync();
+        }
+    }
+
+    private void registerEndpoints() {
+        env.register(remoteLockService);
+        env.register(timestampService);
+        if (config.leader().isPresent()) {
+            localPaxosServices = Leaders.createAndRegisterLocalServices(env, config.leader().get(), userAgent);
+        }
+        env.register(persistentLockService);
+        env.register(new CheckAndSetExceptionMapper());
+        env.register(sweeperService);
+    }
+
+    private void initialiseAsync() {
+        runAsync.accept(() -> {
+            while (uninitialised()) {
+                try {
+                    initialise();
+                } catch (Throwable th) {
+                    log.warn("Async initialisation failed, retrying in 30 seconds.", th);
+                    Uninterruptibles.sleepUninterruptibly(30, TimeUnit.SECONDS);
```

Comment:
this should be a constant; also, 30 seconds might be a bit too long as discussed above - maybe 5 seconds?